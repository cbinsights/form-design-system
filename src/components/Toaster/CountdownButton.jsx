import React from 'react';
import PropTypes from 'prop-types';
import { Transition } from 'react-transition-group';
import FDS from 'lib/dictionary/js/styleConstants';
import IconButton from 'components/IconButton';
import DenyIcon from 'lib/icons/react/DenyIcon';

/**
 *
 * @param {Number} circleSize bounding box size of circle in px
 * @param {Number} strokeWidth width of circle stroke
 */
export const getCircleInfo = (circleSize, strokeWidth) => {
  const radius = (circleSize - strokeWidth) / 2;

  return {
    r: radius,
    c: Math.PI * radius * 2,
    centerOffset: circleSize / 2,
  };
};

const CountdownButton = ({ label = 'Close', Icon = DenyIcon, duration, onClick }) => {
  const circleSize = 34;
  const strokeWidth = '2';
  const { r, c, centerOffset } = getCircleInfo(circleSize, strokeWidth);

  const defaultCircleStyle = {
    // strokeDasharray sets up a dashed line in which each dash is equal to
    // the circumference of the circle.
    //
    // We offset the dash by 0 to start to create a full circle
    strokeDasharray: c,
    strokeDashoffset: 0,
    transition: `all ${duration}ms linear`,

    // start the countdown at "midnight" position of circle
    transformOrigin: 'center center',
    transform: 'rotate(-90deg)',
  };

  const transitionStyles = {
    entering: { strokeDashoffset: 0 },

    // offsetting the dash by the exact circle circumference lines up
    // the empty part of the dashed stroke with the circle circumference
    entered: { strokeDashoffset: c },
  };

  return (
    <div
      className="countdownButton"
      style={{
        width: `${circleSize}px`,
        height: `${circleSize}px`,
      }}
    >
      <div className="countdownButton-button">
        <IconButton onClick={onClick} radius="circle" Icon={Icon} label={label} />
      </div>
      <Transition
        in={true}
        appear
        timeout={{
          appear: 1,
          enter: duration,
        }}
        mountOnEnter
      >
        {(state) => (
          <svg
            className="countdownButton-animation"
            width={circleSize}
            height={circleSize}
            viewBox={`0 0 ${circleSize} ${circleSize}`}
          >
            <circle
              cx={centerOffset}
              cy={centerOffset}
              r={r}
              strokeWidth={strokeWidth}
              fill="transparent"
              stroke={FDS.FONT_COLOR_HINT}
              style={{
                ...defaultCircleStyle,
                ...transitionStyles[state],
              }}
            />
          </svg>
        )}
      </Transition>
    </div>
  );
};

CountdownButton.propTypes = {
  /** Duration in ms of countdown animation */
  duration: PropTypes.number.isRequired,

  /** Used to render a FDS Icon (should only be used for FDS Icons, e.g. Icon={DenyIcon}) */
  Icon: PropTypes.func,

  /** adds a label to IconButton */
  label: PropTypes.string,

  /** Click callback for button */
  onClick: PropTypes.func,
};

export default CountdownButton;
