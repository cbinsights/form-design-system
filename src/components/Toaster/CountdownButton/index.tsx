import React from 'react';
import { Transition } from 'react-transition-group';
import FDS from 'dictionary/js/styleConstants';
import IconButton from 'components/IconButton';
import DenyIcon from 'icons/react/DenyIcon';
import Icon from 'components/Icon';
export interface CircleInfo {
  radius: number;
  circumference: number;
  centerOffset: number;
}

/**
 *
 * @param {Number} circleSize bounding box size of circle in px
 * @param {Number} strokeWidth width of circle stroke
 */
export const getCircleInfo = (circleSize: number, strokeWidth: number): CircleInfo => {
  const radius = (circleSize - strokeWidth) / 2;

  return {
    radius,
    circumference: Math.PI * radius * 2,
    centerOffset: circleSize / 2,
  };
};

export interface CountdownButtonProps {
  /** Duration in ms of countdown animation */
  duration: number;

  /** Used to render a FDS Icon (should only be used for FDS Icons, e.g. Icon={DenyIcon}) */
  Icon?: Icon;

  /** adds a label to IconButton */
  label?: string;

  /** Click callback for button */
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
}

const CountdownButton = ({
  label = 'Close',
  Icon = DenyIcon,
  duration,
  onClick,
}: CountdownButtonProps): JSX.Element => {
  const circleSize = 34;
  const strokeWidth = 2;
  const { radius, circumference, centerOffset } = getCircleInfo(circleSize, strokeWidth);

  const defaultCircleStyle = {
    // strokeDasharray sets up a dashed line in which each dash is equal to
    // the circumference of the circle.
    //
    // We offset the dash by 0 to start to create a full circle
    strokeDasharray: circumference,
    strokeDashoffset: 0,
    transition: `all ${duration}ms linear`,

    // start the countdown at "midnight" position of circle
    transformOrigin: 'center center',
    transform: 'rotate(-90deg)',
  };

  const transitionStyles = {
    entering: { strokeDashoffset: 0 },

    // offsetting the dash by the exact circle circumference lines up
    // the empty part of the dashed stroke with the circle circumference
    entered: { strokeDashoffset: circumference },
    exiting: undefined,
    exited: undefined,
    unmounted: undefined,
  };

  return (
    <div
      className="countdownButton"
      style={{
        width: `${circleSize}px`,
        height: `${circleSize}px`,
      }}
    >
      <div className="countdownButton-button">
        <IconButton onClick={onClick} radius="circle" Icon={Icon} label={label} />
      </div>
      <Transition
        in={true}
        appear
        timeout={{
          appear: 1,
          enter: duration,
        }}
        mountOnEnter
      >
        {(state) => (
          <svg
            className="countdownButton-animation"
            width={circleSize}
            height={circleSize}
            viewBox={`0 0 ${circleSize} ${circleSize}`}
          >
            <circle
              cx={centerOffset}
              cy={centerOffset}
              r={radius}
              strokeWidth={strokeWidth}
              fill="transparent"
              stroke={FDS.FONT_COLOR_HINT}
              style={{
                ...defaultCircleStyle,
                ...transitionStyles[state],
              }}
            />
          </svg>
        )}
      </Transition>
    </div>
  );
};

export default CountdownButton;
