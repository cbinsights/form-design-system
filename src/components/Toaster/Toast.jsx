import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import FDS from 'lib/dictionary/js/styleConstants';
import CheckIcon from 'lib/icons/react/CheckIcon';
import InformationIcon from 'lib/icons/react/InformationIcon';
import DenyIcon from 'lib/icons/react/DenyIcon';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';
import customPropTypes from 'components/util/customPropTypes';
import Button from 'components/Button';
import IconButton from 'components/IconButton';

import CountdownButton from './CountdownButton';

export const TYPES = ['info', 'warn', 'error', 'success', 'progress'];

const Toast = ({
  isAutoDismiss = true,
  dismissDelay = 4000,
  canDismiss = true,
  dismissOnAction = true,
  type = 'info',
  content,
  actionLabel,
  progress,
  onAction,
  dismissToast /* Passed from Toaster */,
}) => {
  const classNames = cx('toast', `toast--${type}`);

  const icon = {
    info: <InformationIcon size="l" color={FDS.COLOR_AQUA} />,
    success: <CheckIcon size="xs" color={FDS.COLOR_WHITE} />,
    warn: (
      <div role="image" className="fontWeight--bold color--white">
        !
      </div>
    ),
    error: (
      <div role="image" className="fontWeight--bold color--white">
        !
      </div>
    ),
  }[type];

  const useContinuousProgressBar = type === 'progress' && !progress;
  const useAmountProgressBar = type === 'progress' && progress;

  const toastBar = (
    <div className="toast-bar">
      {useContinuousProgressBar && (
        <div className="toast-progressBar toast-progressBar--continuous" />
      )}
      {useAmountProgressBar && (
        <div
          className="toast-progressBar toast-progressBar--amount"
          style={{ width: `${progress}%` }}
        />
      )}
    </div>
  );

  const onActionDismiss = (event) => {
    if (dismissOnAction) dismissToast();
    if (onAction) onAction(event);
  };

  return (
    <div
      className={classNames}
      role={type === 'error' ? 'alert' : 'log'}
      aria-relevant="all"
      aria-atomic="true"
    >
      {toastBar}
      <div className="toast-content alignChild--left--center">
        <Flex>
          {icon && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                <div className="toast-icon media--s">{icon}</div>
              </div>
            </FlexItem>
          )}
          <FlexItem>
            <div className="toast-constrainGrowth">{content}</div>
          </FlexItem>
          {actionLabel && (onAction || (dismissToast && dismissOnAction)) && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                <Button
                  onClick={onActionDismiss}
                  data-test="toast-action"
                  theme="ghost"
                  label={actionLabel}
                />
              </div>
            </FlexItem>
          )}
          {canDismiss && type !== 'progress' && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                {isAutoDismiss ? (
                  <CountdownButton
                    onClick={dismissToast}
                    duration={dismissDelay}
                    label="Close"
                  />
                ) : (
                  <IconButton
                    Icon={DenyIcon}
                    onClick={dismissToast}
                    radius="circle"
                    label="Close"
                  />
                )}
              </div>
            </FlexItem>
          )}
        </Flex>
      </div>
    </div>
  );
};

Toast.propTypes = {
  /** JSX Content of Toast */
  content: PropTypes.node.isRequired,

  /** Type of toast */
  type: PropTypes.oneOf(TYPES),

  /** Label for action button */
  actionLabel: PropTypes.string,

  /** Callback for action button click */
  onAction: PropTypes.func,

  /** Time in ms to auto-dismiss toast */
  dismissDelay: PropTypes.number,

  /** Should this toast auto-dismiss itself? */
  isAutoDismiss: PropTypes.bool,

  /* Controls if toast gets dismiss when the ActionButton is clicked */
  dismissOnAction: PropTypes.bool,

  /** Is this toast user-dismissable? */
  canDismiss: PropTypes.bool,

  /** Number 1-100 declaring % progress */
  progress: customPropTypes.range(1, 100),

  /** Passed by Toaster: function to dismiss the toast */
  dismissToast: PropTypes.func,
};

export default Toast;
