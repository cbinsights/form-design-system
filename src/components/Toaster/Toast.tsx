import React from 'react';
import cc from 'classcat';
import FDS from 'dictionary/js/styleConstants';
import CheckIcon from 'icons/react/CheckIcon';
import InformationIcon from 'icons/react/InformationIcon';
import DenyIcon from 'icons/react/DenyIcon';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';
import Button from 'components/Button';
import IconButton from 'components/IconButton';

import CountdownButton from './CountdownButton';

type ToastType = 'info' | 'warn' | 'error' | 'success' | 'progress';

export interface ToastProps {
  /** JSX Content of Toast */
  content: React.ReactNode;

  /** Type of toast */
  type?: ToastType;

  /** Label for action button */
  actionLabel?: string;

  /** Callback for action button click */
  onAction?: React.MouseEventHandler<HTMLButtonElement>;

  /** Time in ms to auto-dismiss toast */
  dismissDelay?: number;

  /** Should this toast auto-dismiss itself? */
  isAutoDismiss?: boolean;

  /* Controls if toast gets dismiss when the ActionButton is clicked */
  dismissOnAction?: boolean;

  /** Is this toast user-dismissable? */
  canDismiss?: boolean;

  /** Number 1-100 declaring % progress */
  progress?: number;

  /** Passed by Toaster: function to dismiss the toast */
  dismissToast?(...args: unknown[]): unknown;
}

const Toast = ({
  isAutoDismiss = true,
  dismissDelay = 4000,
  canDismiss = true,
  dismissOnAction = true,
  type = 'info',
  content,
  actionLabel,
  progress,
  onAction,
  dismissToast /* Passed from Toaster */,
}: ToastProps): JSX.Element => {
  const classNames = cc(['toast', `toast--${type}`]);

  const icon = {
    info: <InformationIcon size="l" color={FDS.COLOR_WHITE} />,
    success: <CheckIcon size="xs" color={FDS.COLOR_WHITE} />,
    warn: (
      <div role="image" className="fontWeight--bold color--white">
        !
      </div>
    ),
    error: (
      <div role="image" className="fontWeight--bold color--white">
        !
      </div>
    ),
    progress: undefined,
  }[type];

  const accentColor = {
    info: FDS.COLOR_BLUE,
    warn: FDS.COLOR_ORANGE,
    error: FDS.COLOR_RED,
    success: FDS.COLOR_GREEN,
    progress: FDS.COLOR_AQUA,
  }[type];

  const iconStyle = { backgroundColor: accentColor };

  const onActionDismiss: React.MouseEventHandler<HTMLButtonElement> = (event) => {
    if (dismissOnAction && dismissToast) dismissToast();
    if (onAction) onAction(event);
  };

  return (
    <div
      className={classNames}
      role={type === 'error' ? 'alert' : 'log'}
      aria-relevant="all"
      aria-atomic="true"
    >
      <div className="toast-bar">
        <svg
          xmlns="http://www.w3.org/2000/svg"
          className={cc([
            'toast-svg',
            { 'toast-svg--animate': type === 'progress' && !progress },
          ])}
        >
          <line
            x1="0%"
            y1="50%"
            x2={type === 'progress' && progress ? `${progress}%` : '100%'}
            y2="50%"
            stroke={accentColor}
            strokeWidth={FDS.SPACE_S}
            strokeDasharray={type === 'progress' && !progress ? '200 400' : ''}
          />
        </svg>
      </div>
      <div className="toast-content alignChild--left--center">
        <Flex>
          {icon && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                <div className="toast-icon media--s" style={iconStyle}>
                  {icon}
                </div>
              </div>
            </FlexItem>
          )}
          <FlexItem>
            <div className="toast-constrainGrowth">{content}</div>
          </FlexItem>
          {actionLabel && (onAction || (dismissToast && dismissOnAction)) && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                <Button
                  onClick={onActionDismiss}
                  data-test="toast-action"
                  theme="ghost"
                  label={actionLabel}
                />
              </div>
            </FlexItem>
          )}
          {canDismiss && type !== 'progress' && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                {isAutoDismiss ? (
                  <CountdownButton
                    onClick={dismissToast}
                    duration={dismissDelay}
                    label="Close"
                  />
                ) : (
                  <IconButton
                    Icon={DenyIcon}
                    onClick={dismissToast}
                    radius="circle"
                    label="Close"
                  />
                )}
              </div>
            </FlexItem>
          )}
        </Flex>
      </div>
    </div>
  );
};

export default Toast;
