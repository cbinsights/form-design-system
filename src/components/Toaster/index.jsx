import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import Toast from './Toast';

export const Toaster = ({
  toastProps = {},
  dismissDelay = 4000,
  isAutoDismiss = true,
  isOpen = false,
  id,
  onDismiss,
}) => {
  useEffect(() => {
    // We need to explicitly check of canDismiss is NOT true, as it has a default
    // parameter of true that this component does not know about
    if (
      isAutoDismiss &&
      toastProps.type !== 'progress' &&
      toastProps.canDismiss !== false
    ) {
      const timer = setTimeout(() => onDismiss(), dismissDelay);

      return function cleanup() {
        clearTimeout(timer);
      };
    }

    return undefined;
  }, [id, isOpen, isAutoDismiss, toastProps.type, toastProps.canDismiss]);

  return (
    <div aria-live="assertive">
      <TransitionGroup>
        {isOpen && (
          <CSSTransition
            key={id}
            appear
            unmountOnExit
            timeout={380}
            classNames="rtgSlideIn"
          >
            <div className="toaster">
              <Toast
                dismissDelay={dismissDelay}
                dismissToast={onDismiss}
                isAutoDismiss={isAutoDismiss}
                content={toastProps.content}
                type={toastProps.type}
                progress={toastProps.progress}
                canDismiss={toastProps.canDismiss}
                actionLabel={toastProps.actionLabel}
                onAction={toastProps.onAction}
                dismissOnAction={toastProps.dismissOnAction}
              />
            </div>
          </CSSTransition>
        )}
      </TransitionGroup>
    </div>
  );
};

const ToasterWrapper = (props) =>
  ReactDOM.createPortal(
    <Toaster {...props} />,
    // eslint-disable-next-line no-undef
    document.body
  );

Toaster.propTypes = {
  /**  A UUID for the toast. Each new toast should have a new UUID. */
  id: PropTypes.string,
  /** Specifies whether Toast is open or not */
  isOpen: PropTypes.bool,
  /** Accepts all props that `Toast` accepts. Refer to Toast component for full documentation */
  toastProps: PropTypes.shape(Toast.propTypes),
  /** Should this toast auto-dismiss itself? */
  isAutoDismiss: PropTypes.bool,
  /** Time in ms to auto-dismiss toast */
  dismissDelay: PropTypes.number,
  /** Callback that is fired when toast is dismissed */
  onDismiss: PropTypes.func,
};

export default ToasterWrapper;
