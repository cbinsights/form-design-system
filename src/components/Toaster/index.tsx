import React, { useEffect } from 'react';
import ReactDOM from 'react-dom';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import Toast, { ToastProps } from './Toast';

const dismissDelay = 4000;

export interface ToasterProps {
  /**  A UUID for the toast. Each new toast should have a new UUID. */
  id?: string;

  /** Specifies whether Toast is open or not */
  isOpen?: boolean;

  /** Should this toast auto-dismiss itself? */
  isAutoDismiss?: boolean;

  onDismiss?: (...args: unknown[]) => void;

  /** Controls the display for your specific toast. */
  toastInstance?: ToastProps;
}

export const Toaster = ({
  isAutoDismiss = true,
  isOpen = false,
  onDismiss,
  toastInstance = {
    content: null,
  },
  id,
}: ToasterProps): JSX.Element => {
  useEffect(() => {
    // We need to explicitly check of canDismiss is NOT true, as it has a default
    // parameter of true that this component does not know about
    if (
      isAutoDismiss &&
      toastInstance.type !== 'progress' &&
      toastInstance.canDismiss !== false &&
      onDismiss
    ) {
      const timer = setTimeout(() => onDismiss(), dismissDelay);

      return function cleanup() {
        clearTimeout(timer);
      };
    }

    return undefined;
  }, [
    id,
    isOpen,
    isAutoDismiss,
    toastInstance.type,
    toastInstance.canDismiss,
    onDismiss,
  ]);

  return (
    <div aria-live="assertive">
      <TransitionGroup>
        {isOpen && (
          <CSSTransition
            key={id}
            appear
            unmountOnExit
            timeout={380}
            classNames="rtgSlideIn"
          >
            <div className="toaster">
              <Toast
                dismissDelay={dismissDelay}
                isAutoDismiss={isAutoDismiss}
                dismissToast={onDismiss}
                content={toastInstance.content}
                type={toastInstance.type}
                progress={toastInstance.progress}
                canDismiss={toastInstance.canDismiss}
                actionLabel={toastInstance.actionLabel}
                onAction={toastInstance.onAction}
                dismissOnAction={toastInstance.dismissOnAction}
              />
            </div>
          </CSSTransition>
        )}
      </TransitionGroup>
    </div>
  );
};

const ToasterWrapper = (props: ToasterProps): React.ReactNode =>
  ReactDOM.createPortal(
    <Toaster {...props} />,
    // eslint-disable-next-line no-undef
    document.body
  );

export default ToasterWrapper;
