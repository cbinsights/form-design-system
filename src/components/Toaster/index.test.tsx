import React from 'react';
import { mount } from 'enzyme';
import Toaster from '.';
import Toast from './Toast';

describe('Toaster component', () => {
  it('matches snapshot', () => {
    const component = mount(
      <Toaster isOpen toastInstance={{ type: 'info', content: <p>test</p> }} />
    );
    expect(component).toMatchSnapshot();
  });

  it('passes dismiss func to toast', () => {
    const onDismiss = jest.fn();
    const component = mount(
      <Toaster
        onDismiss={onDismiss}
        isOpen
        toastInstance={{ type: 'info', content: <p>test</p> }}
      />
    );
    const { dismissToast } = component.find(Toast).props();
    expect(typeof dismissToast).toBe('function');
  });
});
