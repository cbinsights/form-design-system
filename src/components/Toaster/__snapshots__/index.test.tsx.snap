// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toaster component matches snapshot 1`] = `
<ToasterWrapper
  isOpen={true}
  toastInstance={
    Object {
      "content": <p>
        test
      </p>,
      "type": "info",
    }
  }
>
  <Portal
    containerInfo={
      <body>
        <div
          aria-live="assertive"
        >
          <div>
            <div
              class="toaster rtgSlideIn-appear rtgSlideIn-appear-active"
            >
              <div
                aria-atomic="true"
                aria-relevant="all"
                class="toast toast--info"
                role="log"
              >
                <div
                  class="toast-bar"
                >
                  <svg
                    class="toast-svg"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <line
                      stroke="#006699"
                      stroke-dasharray=""
                      stroke-width="8px"
                      x1="0%"
                      x2="100%"
                      y1="50%"
                      y2="50%"
                    />
                  </svg>
                </div>
                <div
                  class="toast-content alignChild--left--center"
                >
                  <div
                    class="flex flex--alignStretch flex--row"
                  >
                    <div
                      class="flexItem--shrink flexItem"
                    >
                      <div
                        class="toast-constrainGrowth alignChild--center--center"
                      >
                        <div
                          class="toast-icon media--s"
                          style="background-color: rgb(0, 102, 153);"
                        >
                          <div
                            class="fds-icon fds-icon--l"
                            style="fill: #ffffff;"
                          >
                            <svg
                              viewBox="0 0 24 24"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <g
                                fill-rule="evenodd"
                              >
                                <g>
                                  <path
                                    d="M12,2 C6.4889971,2 2,6.4889971 2,12 C2,17.511003 6.4889971,22 12,22 C17.511003,22 22,17.511003 22,12 C22,6.4889971 17.511003,2 12,2 Z M12,4 C16.430123,4 20,7.5698774 20,12 C20,16.430123 16.430123,20 12,20 C7.5698774,20 4,16.430123 4,12 C4,7.5698774 7.5698774,4 12,4 Z M11,7 L11,9 L13,9 L13,7 L11,7 Z M11,11 L11,17 L13,17 L13,11 L11,11 Z"
                                  />
                                </g>
                              </g>
                            </svg>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="flexItem"
                    >
                      <div
                        class="toast-constrainGrowth"
                      >
                        <p>
                          test
                        </p>
                      </div>
                    </div>
                    <div
                      class="flexItem--shrink flexItem"
                    >
                      <div
                        class="toast-constrainGrowth alignChild--center--center"
                      >
                        <div
                          class="countdownButton"
                          style="width: 34px; height: 34px;"
                        >
                          <div
                            class="countdownButton-button"
                          >
                            <button
                              class="shape--circle fdsIconButton rounded--all border--focus--noTransition transition--default alignChild--center--center fdsIconButton--ghost fdsIconButton--m"
                              title="Close"
                            >
                              <div
                                class="fdsIconButton--loading-spinnerWrapper"
                              >
                                <div
                                  class=""
                                />
                              </div>
                              <span
                                class=""
                              >
                                <div
                                  class="fds-icon fds-icon--s"
                                  style="width: 18px; height: 18px;"
                                >
                                  <svg
                                    viewBox="0 0 24 24"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <g
                                      fill-rule="evenodd"
                                    >
                                      <g>
                                        <polygon
                                          points="19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12"
                                        />
                                      </g>
                                    </g>
                                  </svg>
                                </div>
                              </span>
                            </button>
                          </div>
                          <svg
                            class="countdownButton-animation"
                            height="34"
                            viewBox="0 0 34 34"
                            width="34"
                          >
                            <circle
                              cx="17"
                              cy="17"
                              fill="transparent"
                              r="16"
                              stroke="rgba(64, 64, 64, 0.38)"
                              stroke-width="2"
                              style="stroke-dasharray: 100.53096491487338; stroke-dashoffset: 0; transition: all 4000ms linear; transform-origin: center center; transform: rotate(-90deg);"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </body>
    }
  >
    <Toaster
      isOpen={true}
      toastInstance={
        Object {
          "content": <p>
            test
          </p>,
          "type": "info",
        }
      }
    >
      <div
        aria-live="assertive"
      >
        <TransitionGroup
          childFactory={[Function]}
          component="div"
        >
          <div>
            <CSSTransition
              appear={true}
              classNames="rtgSlideIn"
              in={true}
              key=".0"
              onExited={[Function]}
              timeout={380}
              unmountOnExit={true}
            >
              <Transition
                appear={true}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={false}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={380}
                unmountOnExit={true}
              >
                <div
                  className="toaster"
                >
                  <Toast
                    content={
                      <p>
                        test
                      </p>
                    }
                    dismissDelay={4000}
                    isAutoDismiss={true}
                    type="info"
                  >
                    <div
                      aria-atomic="true"
                      aria-relevant="all"
                      className="toast toast--info"
                      role="log"
                    >
                      <div
                        className="toast-bar"
                      >
                        <svg
                          className="toast-svg"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <line
                            stroke="#006699"
                            strokeDasharray=""
                            strokeWidth="8px"
                            x1="0%"
                            x2="100%"
                            y1="50%"
                            y2="50%"
                          />
                        </svg>
                      </div>
                      <div
                        className="toast-content alignChild--left--center"
                      >
                        <Flex>
                          <div
                            className="flex flex--alignStretch flex--row"
                          >
                            <FlexItem
                              shrink={true}
                            >
                              <div
                                className="flexItem--shrink flexItem"
                              >
                                <div
                                  className="toast-constrainGrowth alignChild--center--center"
                                >
                                  <div
                                    className="toast-icon media--s"
                                    style={
                                      Object {
                                        "backgroundColor": "#006699",
                                      }
                                    }
                                  >
                                    <InformationIcon
                                      color="#ffffff"
                                      size="l"
                                    >
                                      <div
                                        className="fds-icon fds-icon--l"
                                        style={
                                          Object {
                                            "fill": "#ffffff",
                                            "height": undefined,
                                            "width": undefined,
                                          }
                                        }
                                      >
                                        <svg
                                          viewBox="0 0 24 24"
                                          xmlns="http://www.w3.org/2000/svg"
                                        >
                                          <g
                                            fillRule="evenodd"
                                          >
                                            <g>
                                              <path
                                                d="M12,2 C6.4889971,2 2,6.4889971 2,12 C2,17.511003 6.4889971,22 12,22 C17.511003,22 22,17.511003 22,12 C22,6.4889971 17.511003,2 12,2 Z M12,4 C16.430123,4 20,7.5698774 20,12 C20,16.430123 16.430123,20 12,20 C7.5698774,20 4,16.430123 4,12 C4,7.5698774 7.5698774,4 12,4 Z M11,7 L11,9 L13,9 L13,7 L11,7 Z M11,11 L11,17 L13,17 L13,11 L11,11 Z"
                                              />
                                            </g>
                                          </g>
                                        </svg>
                                      </div>
                                    </InformationIcon>
                                  </div>
                                </div>
                              </div>
                            </FlexItem>
                            <FlexItem>
                              <div
                                className="flexItem"
                              >
                                <div
                                  className="toast-constrainGrowth"
                                >
                                  <p>
                                    test
                                  </p>
                                </div>
                              </div>
                            </FlexItem>
                            <FlexItem
                              shrink={true}
                            >
                              <div
                                className="flexItem--shrink flexItem"
                              >
                                <div
                                  className="toast-constrainGrowth alignChild--center--center"
                                >
                                  <CountdownButton
                                    duration={4000}
                                    label="Close"
                                  >
                                    <div
                                      className="countdownButton"
                                      style={
                                        Object {
                                          "height": "34px",
                                          "width": "34px",
                                        }
                                      }
                                    >
                                      <div
                                        className="countdownButton-button"
                                      >
                                        <IconButton
                                          Icon={[Function]}
                                          label="Close"
                                          radius="circle"
                                        >
                                          <button
                                            className="shape--circle fdsIconButton rounded--all border--focus--noTransition transition--default alignChild--center--center fdsIconButton--ghost fdsIconButton--m"
                                            title="Close"
                                          >
                                            <div
                                              className="fdsIconButton--loading-spinnerWrapper"
                                            >
                                              <div
                                                className=""
                                              />
                                            </div>
                                            <span
                                              className=""
                                            >
                                              <DenyIcon
                                                customSize={18}
                                              >
                                                <div
                                                  className="fds-icon fds-icon--s"
                                                  style={
                                                    Object {
                                                      "fill": undefined,
                                                      "height": "18px",
                                                      "width": "18px",
                                                    }
                                                  }
                                                >
                                                  <svg
                                                    viewBox="0 0 24 24"
                                                    xmlns="http://www.w3.org/2000/svg"
                                                  >
                                                    <g
                                                      fillRule="evenodd"
                                                    >
                                                      <g>
                                                        <polygon
                                                          points="19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12"
                                                        />
                                                      </g>
                                                    </g>
                                                  </svg>
                                                </div>
                                              </DenyIcon>
                                            </span>
                                          </button>
                                        </IconButton>
                                      </div>
                                      <Transition
                                        appear={true}
                                        enter={true}
                                        exit={true}
                                        in={true}
                                        mountOnEnter={true}
                                        onEnter={[Function]}
                                        onEntered={[Function]}
                                        onEntering={[Function]}
                                        onExit={[Function]}
                                        onExited={[Function]}
                                        onExiting={[Function]}
                                        timeout={
                                          Object {
                                            "appear": 1,
                                            "enter": 4000,
                                          }
                                        }
                                        unmountOnExit={false}
                                      >
                                        <svg
                                          className="countdownButton-animation"
                                          height={34}
                                          viewBox="0 0 34 34"
                                          width={34}
                                        >
                                          <circle
                                            cx={17}
                                            cy={17}
                                            fill="transparent"
                                            r={16}
                                            stroke="rgba(64, 64, 64, 0.38)"
                                            strokeWidth={2}
                                            style={
                                              Object {
                                                "strokeDasharray": 100.53096491487338,
                                                "strokeDashoffset": 0,
                                                "transform": "rotate(-90deg)",
                                                "transformOrigin": "center center",
                                                "transition": "all 4000ms linear",
                                              }
                                            }
                                          />
                                        </svg>
                                      </Transition>
                                    </div>
                                  </CountdownButton>
                                </div>
                              </div>
                            </FlexItem>
                          </div>
                        </Flex>
                      </div>
                    </div>
                  </Toast>
                </div>
              </Transition>
            </CSSTransition>
          </div>
        </TransitionGroup>
      </div>
    </Toaster>
  </Portal>
</ToasterWrapper>
`;
